src/Elastic.Apm/DiagnosticSource/DiagnosticInitializer.cs

```diff
@@ -30,24 +36,43 @@ internal DiagnosticInitializer(IApmLogger baseLogger, IEnumerable<IDiagnosticLis
 		public void OnNext(DiagnosticListener value)
 		{
 			var subscribedAny = false;
-			foreach (var listener in _listeners)
+
+			if (_listener != null)
 			{
-				if (value.Name == listener.Name)
+				if (value.Name == _listener.Name)
 				{
-					_sourceSubscription = value.Subscribe(listener);
+					_sourceSubscription = value.Subscribe(_listener);
 					_logger.Debug()
 						?.Log("Subscribed {DiagnosticListenerType} to `{DiagnosticListenerName}' events source",
-							listener.GetType().FullName, value.Name);
+							_listener.GetType().FullName, value.Name);
 					subscribedAny = true;
 				}
 			}
+			else
+			{
+				foreach (var listener in _listeners)
+				{
+					if (value.Name == listener.Name)
+					{
+						_sourceSubscription = value.Subscribe(listener);
+						_logger.Debug()
+							?.Log("Subscribed {DiagnosticListenerType} to `{DiagnosticListenerName}' events source",
+								listener.GetType().FullName, value.Name);
+						subscribedAny = true;
+					}
+				}
+			}
 
 			if (!subscribedAny)
 			{
+				var listeners = _listener != null
```

Comment:
nit: I'd call this `listenerTypes`  or something similar to better reflect what this is.